keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt

```diff
@@ -55,62 +58,65 @@ class ServerGroupHandler(
   override val supportedKind = ResourceKind(
     group = apiVersion.group,
-    singular = "server-group",
-    plural = "server-groups"
-  ) to ServerGroupSpec::class.java
+    singular = "cluster",
+    plural = "clusters"
+  ) to ClusterSpec::class.java
 
-  override suspend fun desired(resource: Resource<ServerGroupSpec>): ServerGroup =
+  override suspend fun desired(resource: Resource<ClusterSpec>): Set<ServerGroup> =
     with(resource.spec) {
-      val (imageId, appVersion) = imageResolver.resolveImageId(resource)
-      ServerGroup(
-        moniker = moniker,
-        location = location,
-        launchConfiguration = launchConfiguration.generateLaunchConfiguration(imageId, appVersion),
-        capacity = capacity,
-        dependencies = dependencies,
-        health = health,
-        scaling = scaling,
-        tags = tags
-      )
+      val resolvedImages = imageResolver.resolveImageId(resource)
+      resolve(resolvedImages)
     }
 
-  override suspend fun current(resource: Resource<ServerGroupSpec>): ServerGroup? =
-    cloudDriverService.getServerGroup(resource)
+  override suspend fun current(resource: Resource<ClusterSpec>): Set<ServerGroup> =
+    cloudDriverService.getServerGroups(resource)
 
   override suspend fun upsert(
-    resource: Resource<ServerGroupSpec>,
-    resourceDiff: ResourceDiff<ServerGroup>
-  ): List<Task> {
-    val spec = resourceDiff.desired
-    val job = when {
-      resourceDiff.isCapacityOnly() -> spec.resizeServerGroupJob(resource.serviceAccount)
-      else -> spec.createServerGroupJob(resource.serviceAccount)
-    }
+    resource: Resource<ClusterSpec>,
+    resourceDiff: ResourceDiff<Set<ServerGroup>>
+  ): List<Task> =
+    resourceDiff
+      .toIndividualDiffs()
+      .filter { diff -> diff.hasChanges() }
+      .map { diff ->
+        val spec = diff.desired
+        val job = when {
+          diff.isCapacityOnly() -> spec.resizeServerGroupJob(resource.serviceAccount)
+          else -> spec.createServerGroupJob(resource.serviceAccount)
+        }
 
-    log.info("Upserting server group using task: {}", job)
-    val description = "Upsert server group ${spec.moniker.name} in ${spec.location.accountName}/${spec.location.region}"
+        log.info("Upserting server group using task: {}", job)
+        val description = "Upsert server group ${spec.moniker.name} in ${spec.location.accountName}/${spec.location.region}"
 
-    return orcaService
-      .orchestrate(
-        resource.serviceAccount,
-        OrchestrationRequest(
-          description,
-          spec.moniker.app,
-          description,
-          listOf(Job(job["type"].toString(), job)),
-          OrchestrationTrigger(resource.id.toString())
-        ))
-      .also { log.info("Started task {} to upsert server group", it.ref) }
-      // TODO: ugleee
-      .let { listOf(Task(id = it.taskId, name = description)) }
-  }
+        // TODO: this needs to be done async
```

Comment:
oh yeah, I should resolve this before the PR is complete. It's not super high priority, though.